src/docs/arch-s3-extension.dox

```diff
@@ -0,0 +1,78 @@
+implementation for a given bucket. The store requires a bucket and an access token to create the
+filesystem. The bucket is expected to be a string argument, which is the bucket name and the
+corresponding region separated by a ';', e.g. "bucket1;ap-southeast-2". The AWS requires an access
+token to authenticate programmatic access to S3. The access token comprises an access key ID and a
+secret access key as a set. The S3 store requires the token to be provided as a single string
+argument of the access key followed by the secret key, separated by a ';'.
+
+- The S3 store also configures a directory on the file system local to the WiredTiger database as a
+cache for the object files. The cache is meant to improve the read latency for the object files
+read after an initial download from the S3.
+
+@section s3_extension_connection S3 Connection class
+
+The S3Connection class represents an active connection to an S3 bucket. It encapsulates and
+implements all the operations the S3 filesystem requires. The instantiation of the class object
+authenticates with the AWS and validates the existence of and access to the configured bucket.
+The S3Connection exposes an API to list the bucket contents filtered by a directory and a prefix,
+check for an object's existence in the bucket, put an object to the cloud, and get the object from
+the cloud. Though not required for the file system's implementation, the class also provides the
+means to delete the objects to clean up artefacts from the internal unit testing.
+
+@section s3_extension_file_system S3_FILE_SYSTEM
+
+WiredTiger supports a database over a custom file system by giving a \c WT_FILE_SYSTEM interface for
+the end-user to implement. The S3 store itself provides a means to configure as many file systems as
+needed, each linked to an S3 bucket. These file systems, called the \c S3_FILE_SYSTEM, are a custom
+implementation provided by the S3 store. The S3 filesystem contains an instance of the S3Connection
+class, hence providing an active connection to the S3 bucket embedded in the file system instance.
+
+The S3 file system uses the functionality provided by the S3 connection to simulate a
+write-once-read-many file system. The filesystem can list the directory contents, check for a
+file's presence, and access the file by creating a filehandle. The files that are not in the cache
+are first downloaded from the bucket and put in the cache before a filehandle is made to access
+them.
+
+The S3 store uses the filesystem's put-object functionality to flush the files into the cloud. The
+Tier Manager inside WiredTiger directs the flush of the files. The files pushed to the cloud are
+also copied to the local cache for easy access.
+
+@section s3_extension_file_handle S3_FILE_HANDLE
+
+The S3 store also implements the \c WT_FILE_HANDLE interface, \c S3_FILE_SYSTEM, to access the files
+on the S3 file system. A handle is created to the copy of the object on the local cache. Since the
```

Comment:
Something about this sentence (`A handle is...`) was hard for me to parse.  How about: "When the handle is created, the local cache gets a copy of the S3 object if it doesn't already have it.  That copy will be accessed by the handle."